### ISSUE TITLE
discv5: NAT traversal via Rendezvous protocol [WIP]

### ISSUE THREAD

**Issue by pipermerriam** [ethereum/devp2p#207]
This issue proposes a mechanism for NAT traversal via UDP hole punching.

This issue borrows from https://github.com/ethereum/portal-network-specs/issues/144 which in-turn borrows from https://blog.ipfs.io/2022-01-20-libp2p-hole-punching/

### Participants

This mechanism involves communication between three nodes:

- Initiator: a node that is behind a NAT, trying to establish a session with the "receiver" node
- Receiver: a node that is behind a NAT
- Rendezvous: a node that is able to communicate with both "initiator" and "receiver"

### Detecting whether you are behind a NAT

Borrowed from: https://twurst.com/articles/stun-without-trust.html#org92b7214

A node in the network should maintain a set `E` which contains all of the `(ip_address, port)` values for outbound packets that have been sent by this node.

When receiving a packet, a node should check whether the packet's `(ip_address, port)` are contained in the set `E`.

- If a node receives a packet such that the `(ip_address, port)` are not in `E` then the node is **not** behind a NAT
- If a node does not receive any packets with `(ip_address, port)` values that are not in `E` within a reasonable amount of time, then the node should assume that they are behind a NAT.

> We suggest 2 minutes as a reasonable amount of time before determining that the node is behind a NAT.

> For practical purposes, an LRU cache should be used to constrain the overall size of the set `E`

### Signalling whether you are behind a NAT

We define a new field in the ENR with the key `"nat"`.

- If the node does not know whether it is behind a NAT, this key should be omitted from the ENR
- If the node is **not** behind a NAT, the value of this key should be set to `0`
- If the node wishes to signal that it **is** behind a NAT, the value of this key should be set to `1`

### Traversing the NAT

We define two new message types:

- RENDEZVOUS
- RENDEZVOUSACK

TODO- finish definition of the protocol and convert this to a PR towards the spec so that people can comment on individual lines.

**Comment by emhane** [comment]
So this means the PING from the "receiver" to the "initiator" is dropped but places the entry in the "receiver's" state table for the "initiator's" PING to the "receiver" to be successful as long as it comes in less than 30 seconds, the timeout of a UDP state table entry in many routers, i.e. the time it takes for the RELAYRESPONSE to reach the "initiator" should be less than 30 seconds? The PONG response of the "receiver" then uses the state table entry that the "initiator's" PING places in its state table to finalise the hole punching?

**Comment by emhane** [comment]
What if the body of the RELAYREQUEST is changed to ```from_node_enr: Enr, to_node_id: NodeId```? Sending the enr of the initiator in the body will supply the receiver with the information it needs to send the PING request to the initiator in step 4.

**Comment by sambacha** [comment]
Why not consider wireguard for this?

**Comment by pipermerriam** [comment]
> What if the body of the RELAYREQUEST is changed to `from_node_enr: Enr, to_node_id: NodeId`? Sending the enr of the initiator in the body will supply the receiver with the information it needs to send the PING request to the initiator in step 4.

Yes, this seems appropriate.  I now see that without this, the "receiver" node will not necessarily have enough information to send the PING to the initiator, which would mean they would end up needing to do a lookup for them in the network to find their ENR.  :+1:

**Comment by emhane** [comment]
I have implemented your protocol outline @pipermerriam with the changes in @AgeManning 's comment above. Furthermore I changed
- [`{response: 2}`](https://github.com/emhane/discv5/blob/5c7662934f79191c7916a1a18f695efeee9a12d3/src/rpc.rs#L19-L21) is in the body of the `RELAYRESPONSE` assembled by the `rendezvous` node in case the request to the `receiver` fails. This means a `RELAYREQUEST` is always responded, or the `initiator` knows the `rendezvous` it chose is faulty.
- The [`nat` field contains the IP address](https://github.com/emhane/discv5/blob/5c7662934f79191c7916a1a18f695efeee9a12d3/src/discv5.rs#L649-L652) and the `nat` field exists 'exclusive or' the `ip` field (this is @AgeManning 's good idea). [The `ip` field has precedence](https://github.com/emhane/discv5/blob/5c7662934f79191c7916a1a18f695efeee9a12d3/src/service.rs#L2206) over the `nat` field if a node messes up its ENR and sets both. If a node is behind an asymmetric NAT, the `udp`/`udp6` field is used to indicate the port at which it can be hole-punched. If the node is behind a symmetric NAT the `udp`/`udp6` field is removed.
- A node, once it has learnt it is behind a NAT, is [responsible for PINGing its peers often enough](https://github.com/emhane/discv5/blob/5c7662934f79191c7916a1a18f695efeee9a12d3/src/service.rs#L2228) to keep the hole punched.
- The NAT traversal protocol is a [feature](https://github.com/emhane/discv5/blob/5c7662934f79191c7916a1a18f695efeee9a12d3/src/discv5.rs#L639) a node can choose to disable (the discv5 topics protocol is also a feature in my implementation for sigpi).

**Comment by emhane** [comment]
I'm changing the `to_node_id` into `to_node_enr` in my implementation because otherwise a node has to store the enr of a peer that is potentially behind a NAT so that it knows where to send the hole-punch-ping upon a RELAYRESPONSE with body `{response: 1}`. This struct to store these ENRs of peers potentially behind a NAT has no obvious capacity limit. It is better if the hole-punch-ping is stateless. The RELAYREQUEST is still small in size in comparison to a NODES response.
